### ISSUE TITLE
ProxyAgent ignores credentials

### ISSUE THREAD

**Issue by negezor** [nodejs/undici#1674]
## Bug Description

I was expecting the classic behavior of passing a proxy URL to be parsed apart. However, ProxyAgent ignores the auth part.

## Reproducible By

```js
import { ProxyAgent, request } from 'undici';

const PROXY_URL = 'http://USERNAME:PASSWORD@HOST';

const dispatcher = new ProxyAgent(PROXY_URL);

const response = await request('https://google.com', {
    dispatcher
});
// HTTP 407 Proxy Authentication Required

console.log('response', response);
```

## Expected Behavior

Server response result

## Logs & Screenshots

```
/home/negezor/projects/repro/node_modules/undici/lib/proxy-agent.js:67
            callback(new RequestAbortedError('Proxy response !== 200 when HTTP Tunneling'))
                     ^

RequestAbortedError [AbortError]: Proxy response !== 200 when HTTP Tunneling
    at Client.connect (/home/negezor/projects/repro/node_modules/undici/lib/proxy-agent.js:67:22)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'UND_ERR_ABORTED'
}
```

## Environment

Node.js v18.10.0
Undici v5.10.0

### Additional context
This is easy to fix, however this is non-obvious behavior.
```js
import { ProxyAgent, request } from 'undici';

const PROXY_URL = 'http://USERNAME:PASSWORD@HOST';

const url = new URL(PROXY_URL);
const dispatcher = new ProxyAgent({
    uri: PROXY_URL,
    auth: Buffer.from(`${url.username}:${url.password}`).toString('base64')
});

const response = await request('https://google.com', {
    dispatcher
});

console.log('response', response);
// Full response
```

**Comment by mcollina** [comment]
cc @RafaelGSS

**Comment by RafaelGSS** [comment]
That's expected. The ProxyAgent uses HTTP Tunneling to perform a request. Would you like to open a PR to clarify it in the documentation?

**Comment by julien-f** [comment]
How is that expected?

Even when using tunneling (using `CONNECT` method), you still have to do a normal HTTP request to the proxy and you can use the `proxy-authenticate` to authenticate against it.

Am I missing anything?

**Comment by metcoder95** [comment]
The `ProxyAgent` parses the URL accordingly to `WHATWG` URL spec, and it respects the `username:password` if no `opts.auth` is provided; do you have a reproducible example?

**Comment by julien-f** [comment]
My bad, it's indeed working correctly, sorry for the noise.

The problem I had was that it does not handle `username` without a `:password`, while it is accepted by Node's `URL` class and appears to be valid if I understand the Wikipedia page:

> userinfo subcomponent [...] may consist of a user name and an optional password preceded by a colon (`:`)
>
> -- https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax

**Comment by Uzlopak** [comment]
Wouldnt it be solved by setting password to an empty string?

**Comment by metcoder95** [comment]
Yeah, indeed won't work as the `ProxyAgent` seems not designed to handle that scenario; but as @Uzlopak said, that's a valid workaround.

The spec for `Basic Auth` scheme does states about separating the `username` and `password` by a colon, being the `password` a possible empty text.

It is up to the server decide what to do on those cases. A PR for supporting can be welcomed from my side

**Comment by julien-f** [comment]
Regardless of the presence of `:`, it does not appear to work natively when the password is empty.

My current work-around:
```js
const uri = new URL(proxy);
const token =
  "Basic " +
  Buffer.from(`${uri.username}:${uri.password}`).toString("base64");
const dispatcher = new ProxyAgent({ token, uri });
```

Unfortunately I don't have time to investigate in `undici` code for a PR.

**Comment by rossilor95** [comment]
Hey folks, can this be closed after   #4935?
